components/state-component.js

```diff
@@ -0,0 +1,49 @@
+// state-component.js
+
+var _ = require('underscore');
+var promiseComponent = require('./promise-component');
+
+/**
+ * Stores the most recent packet for every inPort and calls an onchange function
+ * when they change and after when all required port have received something.
+ * If multiple ports both provide an indexBy property name or function,
+ * only packets that both resolve to the same value with be called together
+ */
+module.exports = function(def){
+    var inPorts = _.isArray(def.inPorts) ? _.object(def.inPorts, []) : def.inPorts;
+    var required = _.compact(_.map(inPorts, function(port, name){
+        return port && port.required && name;
+    }));
+    return promiseComponent(_.defaults({
+        inPorts: _.mapObject(inPorts, function(port, name) {
+            return _.extend({
+                ondata: _.compose(change(required, def.onchange), data(port && port.indexBy, name))
+            }, port);
+        }),
+    }, def));
+};
+
+function data(indexBy, name) {
```

Comment:
indexBy is a property in this module, but would be more flexible as a jsonpointer.